gvc.go

```diff
@@ -133,74 +185,52 @@ func cleanup(path string, opts options) error {
 		if lastVendorPath == "" {
 			lastVendorPath = localPath
 		}
+		lastVendorPathDir := filepath.Dir(lastVendorPath)
 
 		keep := false
+
 		for _, name := range pkgList {
-			// If the file's parent directory is a needed package, keep it.
-			if !info.IsDir() && filepath.Dir(lastVendorPath) == name {
-				if opts.onlyCode {
-					if opts.noTests && strings.HasSuffix(path, "_test.go") {
-						keep = false
-						continue
-					}
-					for _, suffix := range codeSuffixes {
-						if strings.HasSuffix(path, suffix) {
-							keep = true
-						}
-					}
-					// Match keep patterns
-					for _, keepPattern := range opts.keepPatterns {
-						ok, err := doublestar.Match(keepPattern, lastVendorPath)
-						// TODO(sgotti) if a bad pattern is encountered stop here. Actually there's no function to verify a pattern before using it, perhaps just a fake match at the start will work.
-						if err != nil {
-							return fmt.Errorf("bad pattern: %q", keepPattern)
-						}
-						if ok {
-							keep = true
-						}
-					}
-				} else {
-					keep = true
-				}
+			// If a directory is a needed package then keep it
+			keep = keep || info.IsDir() && name == lastVendorPath
+
+			// The remaining tests are only for files
+			if info.IsDir() {
+				continue
 			}
-		}
 
-		// Keep all the legal files inside top repo dir and required package dirs
-		for _, name := range append(repoList, pkgList...) {
-			if !info.IsDir() && filepath.Dir(lastVendorPath) == name {
-				if !opts.noLegalFiles {
-					if IsLegalFile(path) {
-						keep = true
-					}
-				}
+			// Keep legal files in directories that are the parent of a needed package
+			keep = keep || !opts.noLegalFiles && IsLegalFile(path) && strings.HasPrefix(name, lastVendorPathDir)
```

Comment:
peanut gallery: heh, this bit me so often in gps, i made [utility functions](https://github.com/sdboyer/gps/blob/395478daaf6e1f1f54dec560d8bc094ddfc51274/typed_radix.go#L135-L160) for it (now largely subsumed into trie logic)